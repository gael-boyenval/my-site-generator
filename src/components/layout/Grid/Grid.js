import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

import { gutter } from 'Styles/tokens';
import { rem } from 'Styles/utils';

export const gutterWidth = rem(gutter / 2);

const getPredefinedFlexValue = (value) =>
  ({
    fill: { flex: '1 1 0' },
    full: { flex: '100%' },
    default: { flex: 'initial' },
  }[value || 'default']);

const getFixedFlexValue = (value) => ({
  flex: `0 0 ${rem(value)}`,
  width: `${rem(value)}`,
  maxWidth: '100%',
});

const getPercentFlexValue = (value) => ({ flex: `0 0 ${value * 100}%` });

const setCellWidth = (width) => {
  if (width && typeof width === 'string') {
    return getPredefinedFlexValue(width);
  }

  if (width && typeof width === 'number' && width > 1) {
    return getFixedFlexValue(width);
  }

  if (width && typeof width === 'number' && width <= 1) {
    return getPercentFlexValue(width);
  }

  return {
    flex: 'initial',
  };
};

const setPush = (push) => (push && { marginLeft: `${push * 100}%` }) || {};

const GridCell = styled.div.attrs({
  style: ({ size, push }) => Object.assign({}, setCellWidth(size), setPush(push)),
})`
  ${({ alignRight }) =>
    alignRight &&
    css`
      margin-left: auto;
    `};
`;

GridCell.propTypes = {
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  push: PropTypes.number,
};

GridCell.defaultProps = {
  push: null,
};

const Grid = styled.div`
  display: flex;
  flex-flow: row wrap;
  align-items: flex-start;
  flex: 100%;

  ${(props) =>
    props.gutter &&
    css`
      margin-left: -${gutterWidth};
      margin-right: -${gutterWidth};

      > ${GridCell} {
        padding-right: ${gutterWidth};
        padding-left: ${gutterWidth};
      }
    `};

  ${(props) =>
    props.spaceAround &&
    css`
      justify-content: space-around;
    `};

  ${(props) =>
    props.center &&
    css`
      align-items: center;
    `};

  ${(props) =>
    props.inlineNowrap &&
    css`
      display: inline-flex;
      flex-flow: row nowrap;
      word-break: keep-all;

      > ${GridCell} {
        align-self: stretch;
      }
    `};
`;

Grid.Cell = GridCell;

Grid.propTypes = {
  gutter: PropTypes.bool,
  inlineNowrap: PropTypes.bool,
  spaceAround: PropTypes.bool,
  center: PropTypes.bool,
};

export default Grid;
