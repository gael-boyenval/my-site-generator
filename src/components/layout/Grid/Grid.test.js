/* eslint-disable no-undef */
import React from 'react';

import Grid from './Grid';

describe('Component Grid', () => {
  it('should be render without crash', () => {
    const component = shallow(<Grid />);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should be render grid with a cell and a gutter', () => {
    const component = shallow(
      <Grid gutter>
        <Grid.Cell>col 1</Grid.Cell>
        <Grid.Cell>col 2</Grid.Cell>
      </Grid>,
    );

    expect(toJson(component)).toMatchSnapshot();
  });
});
