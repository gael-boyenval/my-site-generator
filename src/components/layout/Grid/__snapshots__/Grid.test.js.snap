// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Grid should be render grid with a cell and a gutter 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex: 100%;
  -ms-flex: 100%;
  flex: 100%;
  margin-left: -3.6rem;
  margin-right: -3.6rem;
}

.c0 > .sc-bdVaJa {
  padding-right: 3.6rem;
  padding-left: 3.6rem;
}

<div
  className="c0"
>
  <styled.div>
    col 1
  </styled.div>
  <styled.div>
    col 2
  </styled.div>
</div>
`;

exports[`Component Grid should be render without crash 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex: 100%;
  -ms-flex: 100%;
  flex: 100%;
}

<div
  className="c0"
/>
`;
